### DISCUSSION TITLE
Using SQL-style LINQ statements

### DISCUSSION THREAD

**Discussion by tkiapril** [planetarium/libplanet#2221] [General]
Currently in our code base we only use functional-style LINQ statements only. As far as we're concerned, there isn't a critical reason why we're not using SQL-style LINQ statements. One of the advantages with SQL-style LINQ statements is that we can use `let` which allows us to allocate an intermediate value for manipulation in the LINQ chain. For example:
```csharp
from currency in delta.TotalSupplyUpdatedCurrencies
where currency.TotalSupplyTrackable
let totalSupply = delta.GetTotalSupplyImpl(currency)
where totalSupply.HasValue
select new KeyValuePair<Currency, FungibleAssetValue>(currency, totalSupply.Value)
```
To do similar things with functional-style LINQ statements, values must be packed into tuples or the statement must be broken into parts with loops and such. However, the syntax of SQL-style LINQ is kind of alienated with the rest of C# syntax, and might require some familiarizing.

For this, I would like to discuss:

- Would we allow SQL-style LINQ statements in our code base, for benefits they bring forth,
- Or should we stick with functional-style LINQ statements to maintain consistency?

**Comment by tkiapril** [comment]
Without much contemplation though, for people without exposure to both SQL and functional programming, I think the learning curve might be comparable (if that's one of the factors to consider.)

**Comment by dahlia** [comment]
For those who don't have idea about two syntaxes of LINQ: <https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/query-syntax-and-method-syntax-in-linq>.

**Comment by longfin** [comment]
> - Would we allow SQL-style LINQ statements in our code base, for benefits they bring forth,
> - Or should we stick with functional-style LINQ statements to maintain consistency?

- AFIK, there is no agreement nor reason for sticking functional-style LINQ. 
- Personally, SQL-style LINQ syntax is still unfamiliar. but it isn't big deal if having benefits.
